File: node/sample_02_index_list.js
/**
 * Sample 2 — Lay danh sach chi so thi truong (Index)
 * =====================================================
 * Hien thi VN-Index, HNX-Index... tren dashboard.
 *
 * Luong:
 *   1. Client goi endpoint indexList (co the loc theo san)
 *   2. API tra ve danh sach chi so va du lieu gia hien tai
 *   3. Map du lieu sang model hien thi UI
 *   4. Lay summary chi tiet cho mot chi so cu the
 *   5. Cache ngan han de giam so lan goi API
 */

import { Auth, Data, Board } from '@ssi.developer/ssi-sdk';
import { config } from './config.js';
import { ensureAuth } from './auth_helper.js';

const auth = new Auth(config);
await ensureAuth(auth);

const data = new Data(auth);

// --- Buoc 1: Lay toan bo chi so ---
const allIndexes = await data.marketData.getIndexes();
console.log(`Tong so chi so: ${allIndexes.length}\n`);

for (const idx of allIndexes) {
  console.log(`  ${idx.index.padEnd(15)} | ${idx.indexName.padEnd(30)} | San: ${idx.board}`);
}

// --- Buoc 2: Loc chi so theo san HOSE ---
console.log('\n--- Chi so san HOSE ---');
const hoseIndexes = await data.marketData.getIndexesByBoard(Board.HOSE);
for (const idx of hoseIndexes) {
  console.log(`  ${idx.index.padEnd(15)} | ${idx.indexName}`);
}

// --- Buoc 3: Lay chi tiet summary cho mot chi so cu the ---
console.log('\n--- VN-Index Summary ---');
const summary = await data.marketData.getIndexSummary('VNINDEX');
if (summary) {
  console.log(`  Gia tri Index     : ${summary.indexValue}`);
  console.log(`  Thay doi          : ${summary.indexChange} (${summary.indexChangePercent}%)`);
  console.log(`  Tong KL khop      : ${summary.totalMatch}`);
  console.log(`  Tong GT khop      : ${summary.totalMatchValue}`);
  console.log(`  Tang / Giam / Dung: ${summary.totalAdvanceStock} / ${summary.totalDeclineStock} / ${summary.totalSteadyStock}`);
  console.log(`  Tran / San        : ${summary.totalCeilingStock} / ${summary.totalFloorStock}`);
} else {
  console.log('Khong lay duoc summary cho VN-Index.');
}

// --- Response Summary ---
console.log(`\n[Response] total_indexes|hose_indexes|first_index`);
console.log(`${allIndexes.length}|${hoseIndexes.length}|${allIndexes.length > 0 ? allIndexes[0].index : ''}`);

File: python/sample_02_index_list.py
"""
Sample 2 — Lấy danh sách chỉ số thị trường (Index)
=====================================================
Hiển thị VN-Index, HNX-Index… trên dashboard.

Luồng:
  1. Client gọi endpoint indexList với filter (exchange, limit, view)
  2. API trả về danh sách chỉ số và dữ liệu giá hiện tại
  3. Map dữ liệu sang model hiển thị UI
  4. Lưu cursor để tải trang kế tiếp khi cần
  5. Cache ngắn hạn để giảm số lần gọi API
"""

from ssi_sdk import Auth, Data
from ssi_sdk.enums import Board
from auth_helper import ensure_auth
from config import config

with Auth(config) as auth:
    ensure_auth(auth)

    with Data(auth) as data:
        # --- Bước 1: Lấy toàn bộ chỉ số ---
        all_indexes = data.market_data.get_indexes()
        print(f"Tổng số chỉ số: {len(all_indexes)}\n")

        for idx in all_indexes:
            print(f"  {idx.index:<15} | {idx.index_name:<30} | Sàn: {idx.board}")

        # --- Bước 2: Lọc chỉ số theo sàn HOSE ---
        print("\n--- Chỉ số sàn HOSE ---")
        hose_indexes = data.market_data.get_indexes_by_board(Board.HOSE)
        for idx in hose_indexes:
            print(f"  {idx.index:<15} | {idx.index_name}")

        # --- Bước 3: Lấy chi tiết summary cho một chỉ số cụ thể ---
        print("\n--- VN-Index Summary ---")
        summary = data.market_data.get_index_summary("VNINDEX")
        if summary:
            print(f"  Giá trị Index    : {summary.index_value}")
            print(f"  Thay đổi         : {summary.index_change} ({summary.index_change_percent}%)")
            print(f"  Tổng KL khớp     : {summary.total_match}")
            print(f"  Tổng GT khớp     : {summary.total_match_value}")
            print(f"  Tăng / Giảm / Đứng: {summary.total_advance_stock} / {summary.total_decline_stock} / {summary.total_steady_stock}")
            print(f"  Trần / Sàn       : {summary.total_ceiling_stock} / {summary.total_floor_stock}")
        else:
            print("Không lấy được summary cho VN-Index.")
